### ISSUE TITLE
[Redfish] rpower implement

### ISSUE THREAD

**Issue by xuweibj** [xcat2/xcat2-task-management#517]
* [x] boston support
    * [x] state
    * [x] bmcstate
    * [x] on/off/reset/boot
    * [x] bmcreboot

* [x] witherspoon support
    * [x] state
    * [x] bmcstate
    * [x] on/off/reset/boot
    * [x] bmcreboot

Chassis
```
# curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: <token>" https://<bmcip>/redfish/v1/Chassis/[Planar|Backplane]
"PowerState": "Off",
```
BMC: ``/redfish/v1/Managers/BMC``
Computer: ``/redfish/v1/Systems/Computer``

Set computer power status:
```
curl -k -X POST -H "Content-Type: application/json" -H "X-Auth-Token: <token>" https://<bmcip>/redfish/v1/Systems/Computer/Actions/ComputerSystem.Reset -d '{"ResetType": "On"}'
```

bmcreboot
```
curl -k -X POST -H "Content-Type: application/json" -H "X-Auth-Token: <token>" https://<bmcip>/redfish/v1/Managers/BMC/Actions/Manager.Reset -d '{"ResetType": "ForceRestart"}'
```

**Comment by whowutwut** [comment]
@xuweibj This description is great in there that shows the interface, but really this task should be to create a UT script that will run the test against the endpoint.   

Then we should have another issue to implement the xCAT related command, but I think the priority should be to create the UT framework because having this automation will save a lot of time in the future, also if we change to support v1 -> v2, it would be good to have the UT cases working first before we start implementing xCAT related functions.

**Comment by robin2008** [comment]
Could we have the same template for this kinds of tasks? @xuweibj 

- Investigation

1. Login the BMC
Note: different server model may have different login URI,  you can get it via :
`curl -k -X GET -H "Content-Type: application/json" https://$BMCIP/redfish/v1/ -s`
The URI should be something like this
```
 ...
    "Links": {
        "Sessions": {
            "@odata.id": "/redfish/v1/SessionService/Sessions"
        }
    },
 ...
```
Refer to login task (xxx) for more details.

```
export BMCIP=boston36-bmc
export LOGIN_URI="/redfish/v1/SessionService/Sessions"
curl -k -X POST -H "Content-Type: application/json" https://${BMCIP}${LOGIN_URI} -d '{"UserName": "ADMIN", "Password": "ADMIN"}' -s -D /tmp/mytoken

#save the token for later using
TOKEN=$(grep X-Auth-Token /tmp/mytoken|awk '{print $2}')
```
2. Get power state of chassis,  URI: `/redfish/v1/Chassis/<id>` id could be `Backplane` or `Planar`
```
export STATE_URI=/redfish/v1/Chassis/Backplane
curl -k -X GET -H "Content-Type: application/json" -H "X-Auth-Token: $TOKEN" https://${BMCIP}${STATE_URI}

{
    "@odata.context": "/redfish/v1/$metadata#Chassis.Chassis",
    "@odata.type": "#Chassis.v1_5_0.Chassis",
    "@odata.id": "/redfish/v1/Chassis/Backplane",
    "Id": "Backplane",
    "Name": "OpenPOWER System Chassis",
    "ChassisType": "RackMount",
    "Manufacturer": "IBM",
    "Model": "NONE",
    "SerialNumber": "1309ALA",
    "PartNumber": "9006-22C",
    "AssetTag": "NONE",
    "PowerState": "On",
    "IndicatorLED": "Off",
    "Status": {
        "State": "Enabled",
        "Health": "OK"
    },
    "PhysicalSecurity": {
        "IntrusionSensorNumber": 226,
        "IntrusionSensor": "HardwareIntrusion",
        "IntrusionSensorReArm": "Manual"
    },
    "Thermal": {
        "@odata.id": "/redfish/v1/Chassis/Backplane/Thermal"
    },
    "Power": {
        "@odata.id": "/redfish/v1/Chassis/Backplane/Power"
    },
    "Links": {
        "ComputerSystems": [
            {
                "@odata.id": "/redfish/v1/Systems/Computer"
            }
        ],
        "ManagedBy": [
            {
                "@odata.id": "/redfish/v1/Managers/BMC"
            }
        ]
    }
}

Just get the state from `PowerState`

```
3. Power on/off
```
foo
```
4. Power reset
...


- UT script (#PR xxx)

- Implementation in the below ticket:
(Create sub ticket for each of the CLIs, and added them into this EPIC)

**Comment by jjohnson42** [comment]
for awareness if it matters, pyghmi has begun redfish support:
https://github.com/openstack/pyghmi/commit/b810b02b1e6a065e1f7304d5f282b33a569e0d60

**Comment by jjohnson42** [comment]
I looked at it and had two downsides:
-Another source of dependencies to track.
-It could not accommodate a custom cert validation scheme.  In confluent, the TLS fingerprint of a node is tracked so that a firmware default certificate can actually be used to validate the peer without having to impose a particular CA infrastructure.  pyghmi has a webclient designed to support custom validation.

On the flip side, the upsides were limited.  The abstraction the library provides seems to be just as complex as the underlying REST structure.  Additionally, we have to wait for both DMTF to get something through and then python-redfish to adopt it (e.g. DMTF supports HttpBoot, but python-redfish does not at the moment), when the underlying protocol is really not at all difficult, so long as the code is written to never assume a URL scheme (which makes redfish slow, unfortunately) everything else is easy enough.

**Comment by robin2008** [comment]
I think openstack community also has its own redfish library, just like https://github.com/openstack/sushy and https://github.com/openstack/python-redfish/tree/master/redfish, and the corresponding ironic driver is there. 

but as we know all library will only cover their upper application (just like Ironic, and xCAT r* CLI), but not full support DMTF, so new function there, the only way is to contribute to that OSS, and we might lost the control.

In fact,  most of our codes is still to handle the CLI logic, instead of the interact with redfish device as the interface is already RESTful.

**Comment by jjohnson42** [comment]
It's hard to imagine a library generic enough to support generic redfish yet specific enough to be useful.  The common parts among redfish implementations are so easy to implement that it hardly needs library help beyond http and json handling.  The parts that differ will cause an open ended client to pretty much expose that difference for the caller to handle.

pyghmi makes sense because I want to look compatible to confluent's current calling of pyghmi, but apart from being compatible with existing calling convention, I'm not sure I'd see a good reason to make a library for 'just redfish'.

**Comment by zet809** [comment]
Yes, @jjohnson42 and @robin2008 , as @xuweibj have investigated in https://github.com/xcat2/xcat2-task-management/issues/563
1. each popular libraries only cover the functions they required most
2. it will be a little hard to commit code or function they don't need, or very slow to be merged in.
3. the request/response RESTful interface is easy enough to call and deal with response directly for different users.
So, I think it is reasonable to write our own redfish support library.
BTW, @jjohnson42 , is the pyghmi owned by you? Can you review and merge it in time? If so, maybe we can commit some function into this repo.

**Comment by xuweibj** [comment]
Based on https://github.com/openbmc/openbmc/issues/3458, the interface for power on witherspoon is the similar with boston, so implement is done.
